### ISSUE TITLE
Classify training, image size 640 forced to 224

### ISSUE THREAD

**Issue by chdre** [ultralytics/ultralytics#288]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Training

### Bug

Trying to set resolution to 640 for classification training by
```
imgsz=640
```
as argument, but sizes are forced to 224. Setting resolution to 641 or something else works fine. Cause:
https://github.com/ultralytics/ultralytics/blob/d0b616e41e827e1ac65ee99da1339e707a984989/ultralytics/yolo/v8/classify/train.py#L40-L42
Removing the code snippet solves the problem. 

### Environment

Not relevant

### Minimal Reproducible Example

```
yolo mode=train task=classify imgsz=640 data=path/to/dataset
```

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by knoppmyth** [comment]
According to the documentation I've seen, the classifier only works at 224. Most classifiers I'm aware of, work at this resolution.

**Comment by chdre** [comment]
> According to the documentation I've seen, the classifier only works at 224. Most classifiers I'm aware of, work at this resolution.

A _very_ quick look at the [classification docs](https://docs.ultralytics.com/tasks/classification/) says otherwise. If the classifier was supposed to work only at 224, why would they allow the user to change the image size in the first place?

**Comment by AyushExel** [comment]
Sorry this is weird typo.. We'll fix it

**Comment by chdre** [comment]
Thought so, thanks :)

**Comment by glenn-jocher** [comment]
@chdre thanks for the bug report. Yes default classification models are trained at 224 but the imgsz argument should allow you to modify this to anything else. We'll push a fix.

**Comment by majnas** [comment]
Evan using latest version [**Ultralytics YOLOv8.0.53**], This problem still exist

```shell
(av) dev@dev:~/avgit/defect_classification/code/ultralytics$ yolo classify train data=/home/dev/avgit/defect_classification/dataset/T2W+PP+DDC2022+SU+AV+VAN#23MXBG-POT-A2XR-CLS model=yolov8x-cls.pt batch=12 project="../../runs/Y8" name="POT_T2W+PP+DDC2022+SU+AV+VAN#23MX-POT-A2XR-CLS_Y8X_i640_b12" imgsz=640
Ultralytics YOLOv8.0.53 🚀 Python-3.8.16 torch-1.13.1+cu116 CUDA:0 (NVIDIA GeForce RTX 3080 Laptop GPU, 7982MiB)
yolo/engine/trainer: task=classify, mode=train, model=yolov8x-cls.pt, data=/home/dev/avgit/defect_classification/dataset/T2W+PP+DDC2022+SU+AV+VAN#23MXBG-POT-A2XR-CLS, epochs=100, patience=50, batch=12, imgsz=640, save=True, save_period=-1, cache=False, device=None, workers=8, project=../../runs/Y8, name=POT_T2W+PP+DDC2022+SU+AV+VAN#23MX-POT-A2XR-CLS_Y8X_i640_b12, exist_ok=False, pretrained=False, optimizer=SGD, verbose=True, seed=0, deterministic=True, single_cls=False, image_weights=False, rect=False, cos_lr=False, close_mosaic=10, resume=False, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True, split=val, save_json=False, save_hybrid=False, conf=None, iou=0.7, max_det=300, half=False, dnn=False, plots=True, source=None, show=False, save_txt=False, save_conf=False, save_crop=False, hide_labels=False, hide_conf=False, vid_stride=1, line_thickness=3, visualize=False, augment=False, agnostic_nms=False, classes=None, retina_masks=False, boxes=True, format=torchscript, keras=False, optimize=False, int8=False, dynamic=False, simplify=False, opset=None, workspace=4, nms=False, lr0=0.01, lrf=0.01, momentum=0.937, weight_decay=0.0005, warmup_epochs=3.0, warmup_momentum=0.8, warmup_bias_lr=0.1, box=7.5, cls=0.5, dfl=1.5, fl_gamma=0.0, label_smoothing=0.0, nbs=64, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, degrees=0.0, translate=0.1, scale=0.5, shear=0.0, perspective=0.0, flipud=0.0, fliplr=0.5, mosaic=1.0, mixup=0.0, copy_paste=0.0, cfg=None, v5loader=False, tracker=botsort.yaml, save_dir=../../runs/Y8/POT_T2W+PP+DDC2022+SU+AV+VAN#23MX-POT-A2XR-CLS_Y8X_i640_b129
Overriding model.yaml nc=1000 with nc=2

                   from  n    params  module                                       arguments                     
  0                  -1  1      2320  ultralytics.nn.modules.Conv                  [3, 80, 3, 2]                 
  1                  -1  1    115520  ultralytics.nn.modules.Conv                  [80, 160, 3, 2]               
  2                  -1  3    436800  ultralytics.nn.modules.C2f                   [160, 160, 3, True]           
  3                  -1  1    461440  ultralytics.nn.modules.Conv                  [160, 320, 3, 2]              
  4                  -1  6   3281920  ultralytics.nn.modules.C2f                   [320, 320, 6, True]           
  5                  -1  1   1844480  ultralytics.nn.modules.Conv                  [320, 640, 3, 2]              
  6                  -1  6  13117440  ultralytics.nn.modules.C2f                   [640, 640, 6, True]           
  7                  -1  1   7375360  ultralytics.nn.modules.Conv                  [640, 1280, 3, 2]             
  8                  -1  3  27865600  ultralytics.nn.modules.C2f                   [1280, 1280, 3, True]         
  9                  -1  1   1643522  ultralytics.nn.modules.Classify              [1280, 2]                     
YOLOv8x-cls summary: 183 layers, 56144402 parameters, 56144402 gradients, 154.3 GFLOPs
AMP: running Automatic Mixed Precision (AMP) checks with YOLOv8n...
AMP: checks passed ✅
optimizer: SGD(lr=0.01) with parameter groups 50 weight(decay=0.0), 51 weight(decay=0.00046875), 51 bias
albumentations: RandomResizedCrop(p=1.0, height=224, width=224, scale=(0.08, 1.0), ratio=(0.75, 1.3333333333333333), interpolation=1), HorizontalFlip(p=0.5), ColorJitter(p=0.5, brightness=[0.6, 1.4], contrast=[0.6, 1.4], saturation=[0.6, 1.4], hue=[0, 0]), Normalize(p=1.0, mean=(0.0, 0.0, 0.0), std=(1.0, 1.0, 1.0), max_pixel_value=255.0), ToTensorV2(always_apply=True, p=1.0, transpose_mask=False)
Image sizes 224 train, 224 val
Using 8 dataloader workers
Logging results to ../../runs/Y8/POT_T2W+PP+DDC2022+SU+AV+VAN#23MX-POT-A2XR-CLS_Y8X_i640_b129
Starting training for 100 epochs...

      Epoch    GPU_mem       loss  Instances       Size
      1/100      1.81G     0.1042         12        224:   1%|          | 34/3801 [00:03<05:39, 11.10it/s]
```

**Comment by toulzx** [comment]
@glenn-jocher @AyushExel 

It seems that the bug mentioned above hasn't been fixed, the imgsz will still be resized to 224 only when 640. Do I misunderstand the meaning of the code?

https://github.com/ultralytics/ultralytics/blob/229119c376070d3c7bc0170646e3651155024d38/ultralytics/yolo/v8/classify/train.py#L44-L45

**Comment by glenn-jocher** [comment]
@toulzx thank you for bringing this to our attention. We apologize for the inconvenience this may have caused you and for the lack of communication on our part.

After reviewing the code snippet you mentioned, we can confirm that the issue is still present in the current version of YOLOv8. We appreciate you taking the time to report this issue and we will work to address it as soon as possible.

In the meantime, to train a classification model at a resolution of 640, we recommend using a different argument than `imgsz`. For example, you could try specifying the resolution in the `model.yaml` file, or using the `--cfg` flag to specify a custom config file for the training process.

Thank you again for your report and for your patience while we work on a solution. If you have any other issues or concerns, please don't hesitate to post them to our GitHub repository or to reach out to us on the Ultralytics Community forum.

**Comment by toulzx** [comment]
@glenn-jocher Thanks for the reply, but I think it‘s necessary to describe and reproduce this question in more detail.

### Brief-Description

I clone the repo, and I begin each training with the following command:
```
!pip uninstall ultralytics                        # Ensured use of local cloned project
!python3 /ultralytics/yolo/v8/classify/train.py   # Already added `ultralytics` as sys path  
```

For convenience, made the following brief:
- The file `default.yaml` refers to `/ultralytics/yolo/cfg/default.yaml`
- Params `model` 和 `imgsz` are from the file `default.yaml`
- The file `yolov8n-cls.yaml` refers to `/ultralytics/models/v8/yolov8-cls.yaml`，I modified the param `nc` and the file name and put it under the path `/ultralytics/`


### Reproduce

#### 1

Params in the file `default.yaml` are not modified:
```
model:
imgsz: 640
```

Training logs are as follows:
![1](https://github.com/ultralytics/ultralytics/assets/57480529/0cb2b767-681b-4862-ba5e-cdf81d5f09e5)

Fine.

#### 2

Modified param `imgsz` in the file `default.yaml` :
```
model: 
imgsz: 480
```

Training logs are as follows:
![2](https://github.com/ultralytics/ultralytics/assets/57480529/4a658250-6a1a-4b23-b97b-2f8f0d6581fc)

Fine.

#### 3
Modified params `model`, `imgsz` in the file `default.yaml` :
```
model: yolov8n-cls.yaml
imgsz: 480
```

Training logs are as follows:
![3](https://github.com/ultralytics/ultralytics/assets/57480529/744293e2-3071-4419-a268-839fdb3b469b)

Fine.

#### 4

Modified param `model` in the file `default.yaml` :
```
model: yolov8n-cls.yaml
imgsz: 640
```

Training logs are as follows:
![4](https://github.com/ultralytics/ultralytics/assets/57480529/698e0e0c-984c-41ab-b57e-bdcb3d3ac5aa)

Param `imgsz` has not been modified in the file `default.py`, but the value has been changed elsewhere, and I guess this is related to this code:
https://github.com/ultralytics/ultralytics/blob/db1c5885d51d640f136adff0ba6b0c6783a9d378/ultralytics/yolo/v8/classify/train.py#L44-L45

### Question

So, what do the codes mentioned above really process? Why does it only work when the model is initialized with a `.yaml` file? Why only images with size 640 are processed?

cc @Laughing-q

**Comment by glenn-jocher** [comment]
@toulzx thank you for taking the time to provide a detailed explanation and reproduction steps for this issue. 

Regarding the code you mentioned, the lines you pointed to (`if opt.imgsz == 640`) check if the image size is 640, and if so, set the `imgsz` value to 224. This is done to improve network performance during training by reducing the size of the image input. 

However, as you've noticed, this behavior only occurs when the model is initialized with a `.yaml` file. This is because the `imgsz` parameter defined in the `.yaml` file takes precedence over any other values that may be specified elsewhere, including in the command line arguments or the `default.yaml` file.

Regarding the behavior you observed where the `imgsz` value is set to 224 even when it is not specified in the `default.yaml` file, this is likely due to the fact that the `opt` object used in the `train.py` script is initialized with default values that include an `imgsz` value of 640. 

We hope this explanation helps clarify the behavior you observed. If you have any further questions or concerns, please don't hesitate to reach out to us or to file a new issue on our GitHub repository. Thank you for your contributions to YOLOv8!

**Comment by toulzx** [comment]
> This is done to improve network performance during training by reducing the size of the image input.

@glenn-jocher Thanks for the reply, I see your point. 
But as this reply https://github.com/ultralytics/ultralytics/issues/2400#issuecomment-1535867996 says, if reducing the size is good for training, shouldn't it work on a range of sizes of inputs, not just those of size 640?

**Comment by Laughing-q** [comment]
@toulzx yes you're right! It is because the following code:
```python
 if self.args.imgsz == 640: 
     self.args.imgsz = 224 
```
Let me explain it for you, the default.yaml is shared among all the tasks(detect, segment, pose and classify) but there's a different is that the first three are trained with `imgsz=640` by default but classify trained with `imgsz=224` by default, so we hardcoded it that way for now. The advantage is that all these tasks are working properly with default values(640 for detect/segment/pose and 224 for classify), but one limitation is that training classification with `imgsz=224` is not allowed.
I know this should be fixed but I suppose the priority is not high for us. 
If you figure out a way which could eliminate the limitation, could you please make a PR? Thanks!

**Comment by toulzx** [comment]
> the default.yaml is shared among all the tasks(detect, segment, pose and classify) but there's a difference is that the first three are trained with imgsz=640 by default but classify trained with imgsz=224 by default, so we hardcoded it that way for classification for now. 

@Laughing-q Thank you very much for your reply! Now I figured it out :) 
The code is somewhat confusing.  _My suggestion is to delete this code, and prompt users to use the recommended `imgsz=224` through official documents and training output logs. (if acceptable)_

**The following is my understanding. I hope it would help friends who have the same question before the problem is fixed in the source code.**
I would like to trace the call of that param `imgsz` to explain it. 

---

when we runs mode `train` in default:
```
!python3 /ultralytics/yolo/v8/classify/train.py   # [local]
# or
YOLO().train()
# or
!yolo classify train
```
it will call
https://github.com/ultralytics/ultralytics/blob/dce4efce48a05e028e6ec430045431c242e52484/ultralytics/yolo/v8/classify/train.py#L166-L167

where `train()` is defined in
https://github.com/ultralytics/ultralytics/blob/dce4efce48a05e028e6ec430045431c242e52484/ultralytics/yolo/engine/trainer.py#L166

then `train()` calls `_setup_train()`, `setup_model()` in turn, but `setup_model` has already been overrided in

https://github.com/ultralytics/ultralytics/blob/dce4efce48a05e028e6ec430045431c242e52484/ultralytics/yolo/v8/classify/train.py#L49-L53

so in the file `train.py`, `setup_model()`  will check the param `model`:
https://github.com/ultralytics/ultralytics/blob/dce4efce48a05e028e6ec430045431c242e52484/ultralytics/yolo/v8/classify/train.py#L58-L65

since we use the default value of `model` it will set `yolov8n-cls.py` as default model, which leads to a call for  `get_model()`. so we will find that the following code is executed:
https://github.com/ultralytics/ultralytics/blob/dce4efce48a05e028e6ec430045431c242e52484/ultralytics/yolo/v8/classify/train.py#L43-L45

But if we initialize the model through `yolov8n-cls.yaml`, `get_model()` will not be called and the code upside will not be executed.

---

**In conclusion,**
If we use the yaml file to initialize the model for training, then it will proceed the value of `imgsz` as we expected.
```
model = YOLO('yolov8n-cls.yaml')
model.train(imgsz=640)
# or
!yolo classify train model=yolov8n-cls.yaml imgsz=640
```
But if we specify the model file end with `.pt`, then when `imgsz=640`, it will be modified to 224 (which is recommended in classify task), we need to comment out the affected code.
```
model = YOLO('yolov8n-cls.pt')
model.train(imgsz=640)
# or
YOLO().train(imgsz=640)
# or
!yolo classify train model=yolov8n-cls.pt  imgsz=640
# or
!yolo classify train imgsz=640
```

cc @chdre @majnas @chandan-wiai

**Comment by glenn-jocher** [comment]
@toulzx thank you for the detailed analysis and explanation. We appreciate your feedback on the `imgsz` parameter and the confusion surrounding its behavior. We will take your suggestions into consideration as we work on improving the user experience in future releases of YOLOv8. 

We also wanted to clarify that the behavior you observed involving the modification of `imgsz` to 224 during training only occurs when using the default `yolov8n-cls.pt` model file, and not when using a `.yaml` model file. This is because the `.yaml` file overrides the default value of `imgsz=224` that is set in the code when using the `.pt` model file. 

Thank you for your contribution to YOLOv8 and please let us know if you have any further questions or concerns.

**Comment by Laughing-q** [comment]
@majnas @chdre @toulzx @knoppmyth ok guys I updated the logic here #2637 . And now we support classification training with `imgsz=640`.
training with `imgsz=224` by default:
```bash
yolo classify train data=imagenet100 model=yolov8n-cls.yaml
```
![image](https://github.com/ultralytics/ultralytics/assets/61612323/44e83b79-d27c-44a8-9bae-a567e990b4fb)

training with `imgsz=640`:
```bash
yolo classify train data=imagenet100 model=yolov8n-cls.yaml imgsz=630
```
![image](https://github.com/ultralytics/ultralytics/assets/61612323/f0da9198-5ecc-442b-9eec-e048b90a53d4)

cc @AyushExel @glenn-jocher
